### PR TITLE
MAKE-1271: return iterator when downloading test results

### PR DESCRIPTION
**Pull request by Kimchelly**
JIRA: https://jira.mongodb.org/browse/MAKE-1271

I basically just refactored the LogIterator interface slightly to support returning the new TestResultsIterator and moved the bucket iteration logic to the iterator.

### CODE REVIEW

File: model/test_results_iterator.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	}()
+
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		i.catcher.Wrapf(err, "reading test result '%s'", i.iter.Item().Name())
+		return false
+	}
+	var result TestResult
+	if err := bson.Unmarshal(data, &result); err != nil {
+		i.catcher.Wrapf(err, "unmarshalling test result '%s'", i.iter.Item().Name())
```
Comment (julianedwards) [inline model/test_results_iterator.go:74]: Close iterator

File: model/test_results_iterator.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+	if i.iter == nil {
+		iter, err := i.bucket.List(ctx, "")
+		if err != nil {
+			i.catcher.Wrap(err, "listing bucket contents")
+			return false
+		}
+		i.iter = iter
+	}
+	if !i.iter.Next(ctx) {
```
Comment (julianedwards) [inline model/test_results_iterator.go:49]: We should probably add and errors from `i.iter` to `i.catcher`

File: model/test_results_iterator.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	"context"
+	"io/ioutil"
+
+	"github.com/evergreen-ci/pail"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+type TestResultsIterator interface {
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply model/test_results_iterator.go:13]: Done.

File: model/test_results_iterator.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		grip.Error(message.WrapError(r.Close(), message.Fields{
+			"message": "could not close bucket reader",
+			"bucket":  i.bucket,
+			"path":    i.iter.Item().Name(),
+		}))
+	}()
+
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		i.catcher.Wrapf(err, "reading test result '%s'", i.iter.Item().Name())
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply model/test_results_iterator.go:69]: Added a close in defer.

File: model/test_results_iterator.go (hunk 1 above)
Comment (Kimchelly) [reply model/test_results_iterator.go:74]: Done.

File: model/test_results_iterator.go (hunk 2 above)
Comment (Kimchelly) [reply model/test_results_iterator.go:49]: I added the errors after `Next()` is called.

File: model/test_results_iterator.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+func (i *testResultsIterator) Exhausted() bool {
+	return i.exhausted
+}
+
+func (i *testResultsIterator) Err() error {
+	return i.catcher.Resolve()
+}
+
+func (i *testResultsIterator) Close() error {
+	i.closed = true
```
Comment (julianedwards) [inline model/test_results_iterator.go:104]: this should close the underlying bucket iterator and return that error, but first check that it is closed

Comment (Kimchelly) [reply model/test_results_iterator.go:104]: `pail.BucketIterator` doesn't have a close method.

Comment (julianedwards) [reply model/test_results_iterator.go:104]: lol right

**Pull request merged**